design/handle-backup-of-volumes-by-resources-filters.md

```diff
@@ -0,0 +1,223 @@
+ 
+The size of each single filter value should limit to 256 bytes in case of an unfriendly long variable assignment.
+ 
+If user defined pv filter rules but used Kopia or Restic to do a backup, the backup will fail in validating the resource filter configuration. Same as the situation if using defined volume filter rules but using CSI or plugins to take volume snapshots.
+ 
+For capacity in `pv` or size in `volume`, the value should include the lower value and upper value concatenated by commas. And it has several combinations below:
+- "0,5Gi" or "0Gi,5Gi" which means capacity or size matches from 0 to 5Gi, including value 0 and value 5Gi
+- ",5Gi" which is equal to "0,5Gi"
+- "5Gi," which means capacity or size matches larger than 5Gi, including value 5Gi
+- "5Gi" which is not supported and will be failed in validating configuration.
+ 
+### Filter Reference
+Currently, resources filters are defined in `BackupSpec` struct, it will be more and more bloated with adding more and more filters which makes the size of `Backup` CR bigger and bigger, so we want to store the resources rules in configmap, and `Backup` CRD reference to current configmap.
+
+the `configmap` would be like this:
+```
+apiVersion: v1
+data:
+ filter.json: |-
+   {
+       "resources": {
+           "include": {
+              ...
+           },
+           "exclude": {
+              ...
+           }
+       },
+       "storage": {
+           "pv": {
+               "include": {
+                  ...
+               },
+               "exclude": {
+                  ...
+               }
+           },
+           "volume": {
+               "include": {
+                  ...
+               },
+               "exclude": {
+                  ...
+               }
+           }
+       }
+   }
+kind: ConfigMap
+metadata:
+ creationTimestamp: "2023-01-16T14:08:12Z"
+ name: cm-backup01
+ namespace: default
+ resourceVersion: "17891025"
+ uid: b73e7f76-fc9e-4e72-8e2e-79db717fe9f1
+```
+
+a new variable `filterConfigmap` would be added into `BackupSpec`, it's value is assigned with current resources filters configmap
+```
+apiVersion: velero.io/v1
+kind: Backup
+metadata:
+  name: backup-1
+  spec:
+    filterConfigmap: cm-backup01
+    ...
+```
+
+The configmap basically equivalent to generated by command `kubectl create cm cm-backup01 --from-file filter.json`
+ 
+The configmap only stores those filters assigned value not the whole resources filters.
+ 
+The name of the configmap is `cm-`+`$BackupName`, and it's in Velero install namespace.
+ 
+#### Life-cycle of resource filter configmap
+- If resource filter configmap is only been imported by Velero backup command with flag `from-file`.
+- If the referenced backup has been deleted, the relevant configmap should be removed.
+- Resource filter configmap should also persist into object storage and could be also synchronized automatically when at startup.
```

Comment:
So for the life-cycle of the configmap, if the backup is removed the configmap may not need to be removed and IMO it's OK one configmap being referenced by different backups.

We should probably upload the referenced configmap to object storage but syncing the configmap is probably unnecessary, b/c the configmap with the same name can have different content when different backups are taken, syncing the configmap map cause one overwrites another, at the sametime it doesn't help in the restore workflow.

Reply:
> one configmap being referenced by different backups
How do we user reference configmap to another different backup? if we want to implement it, should we introduce another flag like `pv-backup-policy-from` when creating a backup to reference another existing backup?

> syncing the configmap is probably unnecessary
agree with you

Reply:
And should we store configmap to object storage if we don't sync it ?